### PR TITLE
SONARJAVA-1296 handle this() and super() constructors

### PR DESCRIPTION
**Pull request by m-g-sonar**

### CODE REVIEW

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 1]
```diff
@@ -391,21 +391,28 @@ private Resolution findMethod(Env env, JavaType callSite, JavaType site, String
 
   private Resolution findMethod(Env env, JavaType callSite, JavaType site, String name, List<JavaType> argTypes, List<JavaType> typeParams, boolean autoboxing) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "findMethod" is 13 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 2]
```diff
@@ -672,6 +679,9 @@ boolean isInheritedIn(JavaSymbol symbol, JavaSymbol.TypeJavaSymbol clazz) {
       case Flags.PUBLIC:
         return true;
       case Flags.PRIVATE:
+        if (symbol.name.equals("<init>")) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Move the "<init>" string literal on the left side of this string comparison. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1132)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Reduce the number of returns of this method 6, down to the maximum allowed 5.](https://github.com/SonarSource/sonar-java/blob/9c6f892cd05c08c2c5c9f376cab72d0e7be7655e/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L677) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1142)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 3 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Reduce the number of returns of this method 6, down to the maximum allowed 5.](https://github.com/SonarSource/sonar-java/blob/0bc4b3487ff879109aeaf18fcbeb164600108cb7/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L677) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1142)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* MAJOR [Reduce the number of returns of this method 6, down to the maximum allowed 5.](https://github.com/SonarSource/sonar-java/blob/a1a0883d6f7922cdab949428287c2139b87d0d68/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L677) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1142)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 3]
```diff
@@ -658,6 +671,10 @@ private boolean isAccessible(Env env, JavaSymbol.TypeJavaSymbol site, JavaSymbol
     }
   }
 
+  static boolean sameOutermostClass(JavaSymbol s1, JavaSymbol s2) {
+    return s1.outermostClass() == s2.outermostClass();
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR Change this comparison to use the equals method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1698)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/d2f32603aceb1c35669f161f41fb545c92786d43/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L711) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/a92f3fcd860d264d5157e397644b3577c4d9e6a4/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L711) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/04b5f2e635fbdc144faf84846c3cd38d016ac0e3/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L711) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 4]
```diff
@@ -391,37 +391,45 @@ private Resolution findMethod(Env env, JavaType callSite, JavaType site, String
...
-          bestSoFar = Resolution.resolution(best);
-          bestSoFar.type = resolveTypeSubstitution(((JavaType.MethodJavaType) best.type).resultType, site);
-          JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) best;
-          bestSoFar.type = handleTypeArguments(typeParams, bestSoFar.type, methodSymbol);
+    String searchedName = name;
+    boolean superConstructor = "super".equals(searchedName);
+    if (superConstructor || "this".equals(searchedName)) {
+      searchedName = "<init>";
+    }
+    boolean searchingConstructor = "<init>".equals(searchedName);
```
Comment (benzonico) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: no need to do the equals.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-squid/src/main/java/org/sonar/java/resolve/Resolve.java [hunk 5]
```diff
@@ -389,7 +389,17 @@ private Resolution findMethod(Env env, JavaType callSite, JavaType site, String
     return findMethod(env, callSite, site, name, argTypes, typeParams, false);
   }
 
+  private Resolution findConstructor(Env env, JavaType site, List<JavaType> argTypes, List<JavaType> typeParams, boolean autoboxing) {
+    return findMethod(env, site, site, "<init>", argTypes, typeParams, autoboxing);
+  }
+
   private Resolution findMethod(Env env, JavaType callSite, JavaType site, String name, List<JavaType> argTypes, List<JavaType> typeParams, boolean autoboxing) {
```
Comment (SonarTech) [inline java-squid/src/main/java/org/sonar/java/resolve/Resolve.java]: MAJOR The Cyclomatic Complexity of this method "findMethod" is 14 which is greater than 10 authorized. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues:
* MAJOR 2 major
* INFO 1 info

Watch the comments in this conversation to review them.
Note: the following issues could not be reported as comments because they are located on lines that are not displayed in this pull request:
* INFO [Complete the task associated to this TODO comment.](https://github.com/SonarSource/sonar-java/blob/7a91eb258b392c7323a612eccd9f8dd2df9a2fe9/java-squid/src/main/java/org/sonar/java/resolve/Resolve.java#L698) [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1135)

**Pull request merged**